primitives/beefy/src/commitment.rs

```diff
@@ -15,27 +15,81 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use codec::{Decode, Encode};
 use sp_std::{cmp, prelude::*};
 
 use crate::{crypto::Signature, ValidatorSetId};
 
+/// Id of different payloads in the [`Commitment`] data
+pub type BeefyPayloadId = [u8; 2];
+
+/// Registry of all known [`BeefyPayloadId`].
+pub mod known_payload_ids {
+	use crate::BeefyPayloadId;
+
+	/// A [`Payload`] identifier for Merkle Mountain Range root hash.
+	///
+	/// Encoded value should contain a [`beefy_primitives::MmrRootHash`] type (i.e. 32-bytes hash).
+	pub const MMR_ROOT_ID: BeefyPayloadId = *b"mh";
+}
+
+/// A BEEFY payload type allowing for future extensibility of adding additional kinds of payloads.
+///
+/// The idea is to store a vector of SCALE-encoded values with an extra identifier.
+/// Identifiers MUST be sorted by the [`BeefyPayloadId`] to allow efficient lookup of expected
+/// value. Duplicated identifiers are disallowed. It's okay for different implementations to only
+/// support a subset of possible values.
+#[derive(Decode, Encode, Debug, PartialEq, Eq, Clone, Ord, PartialOrd, Default, Hash)]
+pub struct Payload(Vec<(BeefyPayloadId, Vec<u8>)>);
+
+impl Payload {
+	/// Returns a raw payload under given `id`. If the `BeefyPayloadId` is not found in the payload
+	/// `None` is returned.
+	pub fn get_raw(&self, id: &BeefyPayloadId) -> Option<&Vec<u8>> {
+		self.0.iter().find_map(
+			|(payload_id, bytes)| {
+				if payload_id == id {
+					Some(bytes)
+				} else {
+					None
+				}
+			},
+		)
+	}
+
+	/// Returns a decoded value `T`, in case the value is not there or the encoding does not match
+	/// it returns `None`.
+	pub fn get_decoded<T: Decode>(&self, id: &BeefyPayloadId) -> Option<T> {
+		self.0.iter().find_map(|(payload_id, bytes)| {
+			if payload_id == id {
+				T::decode(&mut &bytes[..]).ok()
+			} else {
+				None
+			}
+		})
+	}
+
+	/// Push a value that implements [`codec::Encode`] with a given id
+	/// to the back of the payload vec. This method will internally sort the payload vec
+	/// after every push.
+	pub fn push<T: Encode>(&mut self, id: BeefyPayloadId, value: T) {
+		self.0.push((id, value.encode()));
+		self.0.sort_by(|(id_1, _), (id_2, _)| {
+			id_1.partial_cmp(id_2).expect("well_known_payload_ids are always comparable")
```

Comment:
`PartialOrd` + `expect` = `Ord`

```suggestion
			id_1.cmp(id_2)
```